Add `lib/track-element-size`.

sgomes opened the pull request:
The new library provides React hooks to keep track of an element's size in the DOM.
This PR also rewrites the Chart component to use the new library.

This PR doesn't introduce any performance fixes per se, but rather adds building blocks that encapsulate all of the complexity of tracking a component's size in a performant way, making it easier to build performant new components in the future.

#### Changes proposed in this Pull Request

* Add `lib/track-element-size`
* Rewrite `Chart` to make use of `lib/track-element-size`

#### Testing instructions

* Ensure the unit tests for `lib/track-element-size` pass.
* Ensure that the Chart (in Stats) continues to work normally.

sgomes committed:
Add lib/track-element-size.

This new library provides React hooks to keep track of an element's size
in the DOM.

This commit also rewrites the Chart component to use the new library.

sgomes commented:
Rebased on master.

sgomes commented:
Rebased on master.

blowery commented on client/lib/track-element-size/README.md:
```diff
@@ -0,0 +1,90 @@
+
+function SampleComponent() {
+  // Get ref, rect and subscribe to changes.
+  const [ ref, rect ] = useWindowResizeCallback( ref, resizeCallback );
```
should this be `useWindowResizeRect`?

sgomes replied:
Yes it should! And it shouldn't have any parameters. Good catch!

sgomes replied:
Fixed in latest commit.

blowery commented on client/lib/track-element-size/index.js:
```diff
@@ -0,0 +1,92 @@
+ * React hook that subscribes a consumer to changes to the bounding client rect of an element, based
+ * on window resize events.
+ * Does not notify when an element resizes due to reasons other than the window resizing.
+ * Uses throttling on the events, to avoid making changes too often.
```
is there an advantage to throttling to a fixed time window vs throttling to something adaptive, like `requestAnimationFrame`?

sgomes replied:
I generally avoid using `rAF` for anything which shouldn't be locked to screen refresh rate (anything that isn't an animation, really). Besides being too frequent in this case (at just 16ms), it can become even faster in the future if browsers ever adopt >60fps refresh rates in JS, as Safari has already done in CSS.

Resizes already generally break the frame budget, so there's no point adding more work to that and attempting to recalculate this on every frame.

blowery replied:
Thanks for the explanation! Sounds like a good candidate for the development devdocs?

sgomes replied:
This explanation, you mean? Sure! Anywhere in particular you had in mind?

blowery commented on client/lib/track-element-size/index.js:
```diff
@@ -0,0 +1,92 @@
+		// Notify consumer of bounding client rect change.
+		const handleRectChange = rect => {
+			if ( rect ) {
+				const jsonRect = rect.toJSON ? rect.toJSON() : rect;
```
TIL https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect#Notes

Browsers. Sigh.

I assume this toJSON is to convert the DOMRect to a plain object?

sgomes replied:
That's right. It seemed like a saner API to return a JSON object that we assume can have different properties, rather than a `ClientRect` or `DOMRect`, depending on browser implementation.

sgomes committed:
Fix useWindowResizeRect sample.

sgomes committed:
Add typing to lib/track-element-size.

Now with 110% more type safety.

sgomes commented:
Rebased on master and added types 🎉

dmsnell commented on client/lib/track-element-size/README.md:
```diff
@@ -0,0 +1,90 @@
+  );
+
+  // Get ref and subscribe to changes.
+  const ref = useWindowResizeCallback( ref, resizeCallback );
```
I think we didn't intend to pass `ref` into `useWindowResizeCallback()` did we?

sgomes replied:
Great catch, @dmsnell! 👍 Wow, so many mistakes in such a small readme... 😕

sgomes committed:
Fix another typo in track-element-size readme.

sgomes commented:
Hey folks!

This PR has been open for a while now. I'm happy to close it if there's no interest in a generic utility like this, but I'm pinging you in case it's just dropped off your radars. Thanks! 🙂

jsnajdr commented:
Putting it on my radar again 🙂

sirreal commented on client/lib/track-element-size/index.ts:
```diff
@@ -0,0 +1,118 @@
+ * Does not notify when an element resizes due to reasons other than the window resizing.
+ * Uses throttling on the events, to avoid making changes too often.
+ *
+ * @returns {Function} The ref to be set on the consumer component.
```
Is this return doc accurate?

sgomes replied:
Nope. Also unnecessary. Good catch!

sirreal commented on client/lib/track-element-size/index.ts:
```diff
@@ -0,0 +1,118 @@
+
+// Indexable version of DOMRect / ClientRect.
+interface UseWindowResizeRect {
+	[index: string]: number | undefined;
```
This is rather annoying 😞

sgomes replied:
No longer needed, thanks to your suggestion. Thanks!

sgomes commented on client/lib/track-element-size/README.md:
```diff
@@ -0,0 +1,90 @@
+
+function SampleComponent() {
+  // Get ref, rect and subscribe to changes.
+  const [ ref, rect ] = useWindowResizeRect();
```
I don't see why not. Each hook instance gets its own state and callback. It might be inefficient, due to the multiple state updates, but it should be safe.

Is there anything in particular you think could fail here?

jsnajdr replied:
Nothing in particular, just curious 🙂 But even if there is one resize event and one layout reflow, this will lead to two state updates, right? Nothing wrong with that, just checking my understanding.

sgomes replied:
Yes, that's correct 🙂 

If you think that having multiple rects being watched in the same component could be a common use-case, I'm happy to try to find some way of optimising it. Although to be fair, I'm not even sure where I'd start... Hooks are designed to be used multiple times, completely independently, so I suspect it would be hard to break that abstraction and batch updates somehow.

sgomes commented on client/lib/track-element-size/README.md:
````diff
@@ -0,0 +1,90 @@
+```
+
+**Note**: The list of properties in the returned rect may vary between browsers. The safe ones to
+use across all of Calypso's supported browsers are:
````
Yes, exactly. They're not present in IE: https://caniuse.com/#feat=getboundingclientrect

sgomes commented on client/lib/track-element-size/index.ts:
```diff
@@ -0,0 +1,118 @@
+
+// Indexable version of DOMRect / ClientRect.
+interface UseWindowResizeRect {
+	[index: string]: number | undefined;
```
I'm not entirely sure whether this syntax allows for additional properties. That was not the intention; rather, I was looking for a way of indicating that the properties are indexable through `Object.keys`. Is there a better way?

sgomes replied:
I made the checks explicit (instead of using a loop), in order to avoid this issue in the first place.

jsnajdr replied:
What was the issue in the first place? On my machine, TypeScript doesn't seem to mind if I pass any valid object to `Object.keys`. It doesn't require any particular subtype.

sgomes replied:
The issue is array-type indexing:

```ts
interface Test {
  foo: string,
  bar: string
}

const test: Test = { foo: 'a', bar: 'b' };

// This is fine.
const array1 = [ test.foo, test.bar ];
// This is not, because of the `[prop]`, even though those are valid prop values.
const array2 = [ 'foo', 'bar' ].map( prop => test[ prop ] );
```

sirreal replied:
The problem is that `Object.keys` returns `string[]`. TypeScript complains when we try to access properties of a defined interface with an index of type `string` because the key type is a union of the known keys `"top" | "bottom" | …`.

One option is to open up the interface so it accepts arbitrary string keys, the other is to be more strict.

We know the keys of the object will be the keys in the interface, but I think the way TypeScript works makes it impossible to make those types work.

An alternative is to tell TypeScript:

```ts
interface I {  a: string; }

const anObject = { a: "a" };
const anI: I = { a: "a" };

// Not happy with this because `Object.keys` returns `string[]`:
//
// for (const k of Object.keys(anI)) {
//
// If we say it's an array of the keys of the interface, we're all good: 
for (const k of Object.keys(anI) as (keyof I)[]) {
    if ( anI[k] === anObject[k] ) {
        console.log( 'equal' );
    }
}
```

sirreal replied:
> // This is not, because of the `[prop]`, even though those are valid prop values.
> const array2 = [ 'foo', 'bar' ].map( prop => test[ prop ] );

This is the same issue, `[ 'foo', 'bar']` is typed as `string[]`, but if we help out like this, it's OK:

```ts
const array2 = (["foo", "bar"] as (keyof Test)[]).map(prop => test[prop]);
```

sgomes replied:
TIL about `keyof`! This looks like an interesting way of handling similar situations in the future. Thanks, @sirreal !

jsnajdr replied:
I see, I can't just access `obj[ prop ]`, because TypeScript doesn't know the value of `prop` and can't typecheck it. Can be annoying, but makes sense.

sirreal commented on client/lib/track-element-size/index.ts:
```diff
@@ -0,0 +1,118 @@
+		// Notify consumer of bounding client rect change.
+		const handleRectChange = ( rect: DOMRect | ClientRect ) => {
+			if ( rect ) {
+				const jsonRect = 'toJSON' in rect ? rect.toJSON() : rect;
```
I was rather surprised to see that `rect.toJSON()` has a return type `any`. I had trouble finding any documentation on it, do you have any links?

sgomes replied:
After some considered thought, I think I'll drop the JSON conversion altogether. It's not that idiomatic in JS, where duck-typing is king.

sirreal commented on client/lib/track-element-size/index.ts:
```diff
@@ -0,0 +1,118 @@
+				for ( const property of Object.keys( jsonRect ) ) {
+					if (
+						lastRect.current === null ||
+						lastRect.current[ property ] !== jsonRect[ property ]
```
This will perform some redundant checks, won't it? Is it enough to satisfy that a common, explicit set of properties satisfied by both `ClientRect` and `DomRect` are equal?

This prevents the need to make an indexable interface to a type that `DomRect` and `ClientRect` can satisfy.

```ts
		// Notify consumer of bounding client rect change.
		const handleRectChange = ( rect: DOMRect | ClientRect ) => {
			if ( rect ) {
				const jsonRect = 'toJSON' in rect ? rect.toJSON() : rect;
				// Avoid notifying consumer if nothing's changed.
				if (
					lastRect.current === null ||
					lastRect.current.bottom !== jsonRect.bottom ||
					lastRect.current.left !== jsonRect.left ||
					lastRect.current.right !== jsonRect.right ||
					lastRect.current.top !== jsonRect.top
				) {
					lastRect.current = jsonRect;
					return callback( jsonRect );
				}
			}
		};
```

Then we can use this type:

```ts
type UseWindowResizeRect = ReturnType< typeof Element.prototype.getBoundingClientRect >;
```

sgomes replied:
I think that despite the verbosity, this is overall better, yes. Thanks for the suggestion!

jsnajdr replied:
TypeScript has a builtin `ClientRect` type that's exactly the thing returned by `getBoundingClientRect`. Only for historical reasons (ancient IEs and TypeScript originate at the same company 🙂) it's not called `DOMRect`. Can we just use `ClientRect`?

sirreal replied:
The type is `getBoundingClientRect(): ClientRect | DOMRect;`, which is annoying but also correct. Different browsers return different things, ie11 returns `ClientRect`, more modern browsers return `DOMRect`.

We should be able to say we return `ClientRect` because that is a minimal type that should be satisfied, but what would the benefit be?

sirreal reviewed (commented):
Left a few notes that you should feel free to dismiss 🙂

sgomes commented on client/lib/track-element-size/index.ts:
```diff
@@ -0,0 +1,118 @@
+		// Notify consumer of bounding client rect change.
+		const handleRectChange = ( rect: DOMRect | ClientRect ) => {
+			if ( rect ) {
+				const jsonRect = 'toJSON' in rect ? rect.toJSON() : rect;
```
After some considered thought, I think I'll drop the JSON conversion altogether. It's not that idiomatic in JS, where duck-typing is king.

jsnajdr replied:
Yep, I think that just returning the native `DOMRect` is a good idea. We don't sanitize DOM elements either, although they differ between browsers.

sgomes committed:
Return the actual Rects, not a plain object.

There's no need to perform the conversion. It's non-idiomatic and
doesn't bring any additional benefits, particularly with explicit types.

sgomes commented on client/lib/track-element-size/index.ts:
```diff
@@ -0,0 +1,118 @@
+				for ( const property of Object.keys( jsonRect ) ) {
+					if (
+						lastRect.current === null ||
+						lastRect.current[ property ] !== jsonRect[ property ]
```
Neatly avoided by making all the comparisons explicit 🙂

jsnajdr replied:
It would be still a good idea to extract the comparison to a `rectIsEqual` function. It's also possible that the `rect` becomes `null` in the lifetime of the component. Consider an element that's optionally rendered:
```jsx
function Chart( { showLegend } ) {
  const [ ref, rect ] = useWindowResizeRect();
  return (
    <div>
      <ChartLine />
      { showLegend && <div ref={ ref }>Legend</div> }
    </div>
  );
}
```

sgomes commented:
Thank you for your excellent comments, @jsnajdr and @sirreal !

I moved away from the conversion to plain objects. This is JS, we do duck-typing around here! 😁 

I also implemented @sirreal's suggestion of explicitly comparing all required properties, instead of working around TS to allow for a loop. It's not *that* verbose, in the end.

Do you agree with these changes?

jsnajdr commented on client/lib/track-element-size/index.ts:
```diff
@@ -0,0 +1,109 @@
+	callback: ( boundingClientRect: UseWindowResizeRect ) => any
+) {
+	const lastRect = useRef< UseWindowResizeRect | null >( null );
+	const [ element, setElement ] = useState< Element | null >( null );
```
Setting state inside a ref callback is not a good idea: the rerender it causes is always redundant. We can store the element in a ref:
```js
const elementRef = useRef< Element | null >( null );
```
and return it (it has type `React.MutableRefObject< Element | null >`) instead of the `callbackRef`.

The custom typedef for `UseWindowResizeCallback` can be completely avoided. All variables have standard React and DOM types.

sgomes replied:
Thank you for making these changes in your commit! 👍

jsnajdr commented on client/lib/track-element-size/index.ts:
```diff
@@ -0,0 +1,109 @@
+	useEffect( () => {
+		// Notify consumer of bounding client rect change.
+		const handleRectChange = ( rect: UseWindowResizeRect ) => {
+			if ( rect ) {
```
The `rect` can become `null` if the watched element disappears. Then we still want to call the callback with a `null` param.

sgomes replied:
Thank you for making these changes in your commit! 👍

jsnajdr commented on client/lib/track-element-size/index.ts:
```diff
@@ -0,0 +1,109 @@
+		// Measure the element in the DOM.
+		// Uses `afterLayoutFlush` to avoid causing layout thrashing.
+		const measureElement = afterLayoutFlush( () => {
+			if ( element && element.getBoundingClientRect ) {
```
`getBoundingClientRect` exists everywhere since IE4, no need for this check.

sgomes replied:
Thank you for making these changes in your commit! 👍

jsnajdr commented on client/lib/track-element-size/index.ts:
```diff
@@ -0,0 +1,109 @@
+		const throttledMeasureElement = throttle( measureElement, THROTTLE_RATE );
+
+		// Set up subscription.
+		if ( element && callback ) {
```
We can do:
```js
if ( ! callback ) {
  return;
}
```
at the very beginning of the `useEffect` hook. Nothing will happen at all (no measuring, no subscription) if callback is not specified.

And by the way, the type of `useWindowResizeCallback` doesn't really allow `null` callback param, does it? We can avoid the check altogether.

sgomes replied:
Right, this was probably a leftover from before I rewrote this in TypeScript :)

sgomes replied:
Also, thank you for making these changes in your commit! 👍

jsnajdr commented on client/lib/track-element-size/index.ts:
```diff
@@ -0,0 +1,109 @@
+	);
+	const callbackRef = useWindowResizeCallback( callback );
+
+	return [ callbackRef, rect ];
```
`setRect` from `useState` is guaranteed to be the same identical function on every call. The `useCallback` memoization is not neccessary for this native low-level function. The whole hook can be this simple:
```js
function useWindowResizeRect(): [ MutableRefObject< Element | null >, ClientRect | null ] {
  const [ rect, setRect ] = useState< ClientRect | null >( null );
  const callbackRef = useWindowResizeCallback( setRect );
  return [ callbackRef, rect ];
}
```

sgomes replied:
Right, since our callback is just to call `setRect` with the same parameters, we can just use `setRect` directly instead of having a memoised arrow function that calls it 👍

sgomes replied:
Thank you for making these changes in your commit! 👍

jsnajdr committed:
Refactor the useWindowResize implementation

jsnajdr committed:
Improve the way how tests handle throttling

jsnajdr commented on client/components/chart/index.jsx:
```diff
@@ -23,196 +23,179 @@ import BarContainer from './bar-container';
+	minBarWidth,
+	minTouchBarWidth,
+	numberFormat,
+	translate,
```
It would be nice to put the `useTranslate` hook here. The component is already hooky... :wink:

jsnajdr replied:
Oh, we also use the `numberFormat` function from the `localized` HOC 🙁 That's not ready as a hook yet.

jsnajdr commented:
@sgomes I pushed my update to the `useWindowResize` hooks as we discussed on video chat earlier today.

The tests needed a little update regarding number of renders and number of `getBoundingClientRect` calls.

I removed your mock for `_.throttle`, because it behaves incorrectly: the first call to `_.throttle` is supposed to happen synchronously. Only the subsequent calls that are going to be delayed will be in `setTimeout` callback.

Instead of that mock, I converted the usage of Jest mock timers to Sinon's `useFakeTimers`. The advantage is that Sinon mocks the `Date.now()` function, too. `_.throttle` calls `Date.now()` internally and doesn't work unless properly mocked.

I already used the same approach some time ago in [tests of debounced Redux state persistence](https://github.com/Automattic/wp-calypso/blob/31337c3a8049423abb234fc72ee4e5fab3519d88/client/state/test/initial-state.js#L394-L395).

jsnajdr reviewed (approved):
I tested the `Chart` changes, too, and I think this PR is ready to :ship:

sgomes commented:
Woohoo! Thank you so much for the review and your own extensive work in improving these hooks, @jsnajdr ! Hopefully they will serve as a nice building block for many components to come :)
